packages/linter/src/utils/rules-requiring-type-checking.ts

```diff
@@ -0,0 +1,89 @@
+import type { Linter } from 'eslint';
+
+// Cache the resolved rules from node_modules
+let knownRulesRequiringTypeChecking: string[] | null = null;
+
+function resolveKnownRulesRequiringTypeChecking(): string[] | null {
+  if (knownRulesRequiringTypeChecking) {
+    return knownRulesRequiringTypeChecking;
+  }
+  try {
+    const { rules } = require('@typescript-eslint/eslint-plugin');
+    const rulesRequiringTypeInfo = Object.entries(rules)
+      .map(([ruleName, config]) => {
+        if ((config as any).meta?.docs?.requiresTypeChecking) {
+          return `@typescript-eslint/${ruleName}`;
+        }
+        return null;
+      })
+      .filter(Boolean);
+    return rulesRequiringTypeInfo;
+  } catch (err) {
+    console.log(err);
+    return null;
+  }
+}
+
+export function hasRulesRequiringTypeChecking(
+  eslintConfig: Linter.Config
+): boolean {
+  knownRulesRequiringTypeChecking = resolveKnownRulesRequiringTypeChecking();
+  if (!knownRulesRequiringTypeChecking) {
+    /**
+     * If (unexpectedly) known rules requiring type checking could not be resolved,
+     * default to assuming that the rules are in use to align most closely with Nx
+     * ESLint configs to date.
+     */
+    return true;
+  }
+  const allRulesInConfig = getAllRulesInConfig(eslintConfig);
+  return allRulesInConfig.some((rule) =>
+    knownRulesRequiringTypeChecking.includes(rule)
+  );
+}
+
+export function removeParserOptionsProjectIfNotRequired(
+  json: Linter.Config
+): Linter.Config {
+  // At least one rule requiring type-checking is in use, do not migrate the config
+  if (hasRulesRequiringTypeChecking(json)) {
+    return json;
+  }
+  removeProjectParserOptionFromConfig(json);
+  return json;
+}
+
+function getAllRulesInConfig(json: Linter.Config): string[] {
+  let allRules = [];
+  if (json.rules) {
+    allRules = [...allRules, ...Object.keys(json.rules)];
+  }
+  if (json.overrides?.length > 0) {
+    for (const override of json.overrides) {
```

Comment:
Use `forEach` instead of `for...of`, as it's more performant and often easier to read.